### PR TITLE
Container for kSNP4 (upgrade from kSNP3)

### PR DESCRIPTION
**Pull request by kissake**
This is a new container for kSNP version 4.0, kSNP4.

Changes include:
 - Faster
 - Clearer status updates in output
 - Better integration with NCBI

Comments:
 - Chose not to place 4.0 under ksnp3 because the major version number was included in the name, and because the updated software is executed as kSNP4.
 -  Chose to include RAxML because it was included in the kSNP3 container; unsure why, but if people want it, it seems better to include it.

Pull Request (PR) checklist:
- [X] Include a description of what is in this pull request in this message.
- [X] The dockerfile successfully builds to a test target for the user creating the PR. (i.e. `docker build --tag samtools:1.15test --target test docker-builds/samtools/1.15` )
- [X] Directory structure as name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. `spades/3.12.0/Dockerfile`)
   - [X] (optional) All test files are located in same directory as the Dockerfile (i.e. `shigatyper/2.0.1/test.sh`)
- [X] Create a simple container-specific [README.md](https://github.com/StaPH-B/docker-builds/blob/master/.github/workflow-templates/readme-template.md) in the same directory as the Dockerfile (i.e. `spades/3.12.0/README.md`)
   - [X] If this README is longer than 30 lines, there is an explanation as to why more detail was needed
- [X] Dockerfile includes the recommended [LABELS](https://github.com/StaPH-B/docker-builds/blob/master/dockerfile-template/Dockerfile#L8-L18) 
- [X] Main [README.md](https://github.com/StaPH-B/docker-builds/blob/master/README.md) has been updated to include the tool and/or version of the dockerfile(s) in this PR
- [X] [Program_Licenses.md](https://github.com/StaPH-B/docker-builds/blob/master/Program_Licenses.md) contains the tool(s) used in this PR and has been updated for any missing

### CODE REVIEW

**Comment by kapsakcj** [comment]
Cool, thanks for the PR! I didn't realize there was a version 4 for kSNP!

Sometime next week I'll have time to do more in depth review.

I went ahead and approved the GH Actions workflow, so that should run shortly to test out the docker image build process

**Comment by erinyoung** [comment]
> Chose to include RAxML because it was included in the kSNP3 container; unsure why, but if people want it, it seems better to include it.

Good idea

File: ksnp4/4.0/Dockerfile
Comment (erinyoung) [inline ksnp4/4.0/Dockerfile:31]: These lines are the "installation lines". The lines that you have are basically the same ones from version 3.1, which was 2 years ago. We've learned a lot since then.

Most of the dockerfiles in this repo split installations into different 'RUN' layers. I've tried to split it out, but there may in a typo in here somewhere.
```
RUN apt-get update && apt-get -y --no-install-recommends install \
  python2 \
  wget \
  tcsh \
  unzip \
  make \
  gcc \
  build-essential && \
  apt-get autoclean && rm -rf /var/lib/apt/lists/*

# install RAxML
RUN  wget https://github.com/stamatak/standard-RAxML/archive/v${RAXMLVER}.tar.gz && \
  tar xvfz v${RAXMLVER}.tar.gz && \
  cd /standard-RAxML-${RAXMLVER} && \
  make -f Makefile.SSE3.PTHREADS.gcc && \
  rm *.o && \
  ln -s $(pwd)/raxmlHPC-PTHREADS-SSE3 /usr/bin/raxml && \
  chmod a+x ./usefulScripts/*.* && \
  sed -i 's/^\$raxmlExecutable =*.*/\$raxmlExecutable = "raxml";/' ./usefulScripts/ProteinModelSelection.pl

# install kSNP4
RUN wget "https://sourceforge.net/projects/ksnp/files/kSNP4 Linux package.zip" && \
  unzip "kSNP4 Linux package.zip" && \
  mv "kSNP4 Linux package/kSNP4pkg/" /usr/local/ && \
  rm -r "kSNP4 Linux package.zip" && \
  mkdir /data
 ```

This helps readability.

Comment (erinyoung) [reply ksnp4/4.0/Dockerfile:31]: Also, does the raxml in apt-get work for kSNP4? (It might make things easier)

```
RUN apt-get update && apt-get -y --no-install-recommends install \
  python2 \
  wget \
  tcsh \
  unzip \
  make \
  gcc \
  raxml \
  build-essential && \
  apt-get autoclean && rm -rf /var/lib/apt/lists/*
  ```

File: ksnp4/4.0/Dockerfile
Comment (erinyoung) [inline ksnp4/4.0/Dockerfile:39]: I saw that RunExamples.sh is in the github account. Does the following work anywhere? (I'd hate for you to have to support two versions of the same file)

```
RUN wget https://raw.githubusercontent.com/kissake/kSNP4/%40/RunExamples.sh && \
bash RunExamples.sh
```

File: ksnp4/4.0/kSNP_test.sh
Comment (erinyoung) [inline ksnp4/4.0/kSNP_test.sh:6]: Notes to self as to what this test does : 

This command downloads some examples files from sourceforge, renames the file to Examples.zip, decompresses that file, and then runs `RunExamples.sh`

I think this is a great test.

**Comment by erinyoung** [comment]
The test looked like worked:
```
#10 97.37 fasta_list: /data/TestResults/in_list1
#10 97.60 Output directory: /data/TestResults/Run13
#10 97.83 annotate_list: /data/Examples/Example1/Results/annotatedGenomes
#10 97.93 Starting kSNP
#10 97.93 Thu Nov  3 16:36:34 UTC 2022
#10 97.93 Configuration for this run:
#10 97.93 input fasta_list: /data/TestResults/in_list1
#10 97.93 output / working directory: /data/TestResults/Run13
#10 97.93 k=13
#10 97.93 annotate_list file: /data/Examples/Example1/Results/annotatedGenomes
#10 97.93 min_fraction_with_locus: 0.75
#10 97.93 Maximum number of open files: 10240
#10 136.4 [00:00:39] (8/10) Getting annotation data from NCBI (8/10) EEE_NorthAmerican_antigenic_variety
#10 138.8 [00:00:41] (8/10) Getting annotation data from NCBI (9/10) EEE_PE-3_0815
#10 141.4 [00:00:44] (8/10) Getting annotation data from NCBI (10/10) EEE_PE6
#10 144.1 [00:00:46] (8/10) Getting annotation data from NCBI (10/10) Done
#10 144.4 [00:00:47] (9/10) Running ParAnn to apply annotation data to SNPs
#10 148.1 [00:00:50] (10/10) Cleanup (0/1) Starting
#10 148.4 [00:00:51] (10/10) Cleanup (1/1) Done
#10 148.4 
#10 148.4 Time used: 00:00:51
#10 148.6 Generated in_list for example2, run 1:
#10 148.6 /data/Examples/Example2/Genomes/ERR579925.fasta	ERR579925
#10 148.6 /data/Examples/Example2/Genomes/VcMS6.fa	VcMS6
#10 148.6 /data/Examples/Example2/Genomes/VcO1-ElTorN16961.fa	VcO1-ElTorN16961
#10 148.6 /data/Examples/Example2/Genomes/VcLMA3984-4.fa	VcLMA3984-4
#10 148.6 /data/Examples/Example2/Genomes/Vc523-80.fa	Vc523-80
#10 148.6 /data/Examples/Example2/Genomes/Vc63-93_MO45.fasta	Vc63-93_MO45
#10 148.9 fasta_list: /data/TestResults/in_list2
#10 149.1 Output directory: /data/TestResults/Run21
#10 149.3 annotate_list: /data/Examples/Example2/Results/annotatedGenomes
#10 149.4 Starting kSNP
#10 149.4 Thu Nov  3 16:37:26 UTC 2022
#10 149.4 Configuration for this run:
#10 149.4 input fasta_list: /data/TestResults/in_list2
#10 149.4 output / working directory: /data/TestResults/Run21
#10 149.4 k=21
#10 149.4 annotate_list file: /data/Examples/Example2/Results/annotatedGenomes
#10 149.4 Maximum number of open files: 10240
#10 149.4 Automatically determining number of CPUs to use:
#10 149.4 The operating system is Linux
#10 149.4 Discovered 2
#10 149.4 Number CPUs: 2
#10 149.4 Python version to be used:
#10 149.4 Python 2.7.18
#10 149.4 SNPS_ALL_INPUT=
#10 149.4 GENBANKFILEINPUT=
#10 149.4 CORE=0
#10 149.4 ML=0
#10 149.4 NJ=0
#10 149.4 VCF=0
#10 149.4 DEBUG=0
#10 149.6 [00:00:00] (1/10) Preprocessing
#10 149.9 [00:00:00] (2/10) Running jellyfish to get kmer counts
#10 249.2 [00:01:40] (3/10) Partitioning kmer counts for parallelization
#10 258.9 [00:01:49] (4/10) Finding SNPs (1/4) Generating SNPs within partitions
#10 285.2 [00:02:16] (4/10) Finding SNPs (2/4) Combining partitions into FASTA files for mummer
#10 285.5 [00:02:16] (4/10) Finding SNPs (3/4) Combining partitions into complete list of SNPs found
#10 285.8 [00:02:16] (4/10) Finding SNPs (4/4) Done finding SNPs
#10 286.1 [00:02:16] (5/10) Running mummer to find SNP positions for annotation (0/3) Starting
#10 286.3 [00:02:17] (5/10) Running mummer to find SNP positions for annotation (1/3) VcMS6
#10 288.6 [00:02:19] (5/10) Running mummer to find SNP positions for annotation (2/3) VcLMA3984-4
#10 290.7 [00:02:21] (5/10) Running mummer to find SNP positions for annotation (3/3) VcO1-ElTorN16961
#10 293.0 [00:02:23] (5/10) Running mummer to find SNP positions for annotation (3/3) Done
#10 293.2 [00:02:23] (6/10) Processing SNPs data (1/4) Adding position data to all_SNPs
#10 293.5 [00:02:24] (6/10) Processing SNPs data (2/4) Sorting SNPs
#10 293.8 [00:02:24] (6/10) Processing SNPs data (3/4) Number and label SNPs
#10 295.9 [00:02:26] (6/10) Processing SNPs data (4/4) Creating VCF file.
#10 296.1 [00:02:26] (7/10) Generating phylogenic trees (0/2) Setup
#10 296.3 [00:02:27] (7/10) Generating phylogenic trees (1/2) parsimony
#10 310.1 [00:02:40] (7/10) Generating phylogenic trees (2/2) Generating unresolved clusters
#10 310.5 [00:02:41] (7/10) Generating phylogenic trees (2/2) Done
#10 310.7 [00:02:41] (8/10) Getting annotation data from NCBI (0/3) Starting
#10 311.0 [00:02:41] (8/10) Getting annotation data from NCBI (1/3) VcMS6
#10 340.3 [00:03:11] (8/10) Getting annotation data from NCBI (2/3) VcLMA3984-4
#10 348.3 [00:03:19] (8/10) Getting annotation data from NCBI (3/3) VcO1-ElTorN16961
#10 354.0 [00:03:24] (8/10) Getting annotation data from NCBI (3/3) Done
#10 354.2 [00:03:24] (9/10) Running ParAnn to apply annotation data to SNPs
#10 397.2 [00:04:07] (10/10) Cleanup (0/1) Starting
#10 397.5 [00:04:08] (10/10) Cleanup (1/1) Done
#10 397.5 
#10 397.7 Time used: 00:04:08
#10 397.7 Waiting for all examples to complete.
#10 397.7 Exiting for review: TestResults
```

**Comment by erinyoung** [comment]
I think this is off to a great start!

This dockerfile was inspired by the dockerfile for a prior version (https://github.com/StaPH-B/docker-builds/blob/master/ksnp3/3.1/Dockerfile) which isn't inherently a bad idea. 

We, however, have learned a lot about maintainability and community involvement in the two years since ksnp3's dockerfile creation.

Things that need to happen for this dockerfile:
1. Create a variable for all relevant versions with `ARG` near the top of the dockerfile, and then replace all instances of the version with the variable. This makes it easier to update later.
2. Add `--no-install-recommends` to your `apt-get install` command and follow with `apt-get autoclean && rm -rf /var/lib/apt/lists/*`. This reduces container size.
3. Separate the installation steps. This helps people like me easily see what's getting installed where. (Yes, I'm aware this adds more layers)
4. Add your handle to the bottom of the main readme as a contributor

You can use the samtools or bbtools dockerfile for an example : https://github.com/StaPH-B/docker-builds/blob/master/samtools/1.16/Dockerfile or https://github.com/StaPH-B/docker-builds/blob/master/bbtools/39.01/Dockerfile

Optional things:
- You can download your RunExamples.sh file from github if it's easier
- I'll let you install raxml with apt-get if it continues to work

File: README.md
Comment (erinyoung) [inline README.md:86]: Please add your handle to the bottom of this file under Authors/Maintainers.

**Comment by erinyoung** [comment]
I've created an empty repository on [quay](https://quay.io/repository/staphb/ksnp4), so once these changes are incorporated, we can get this up there.

**Comment by kissake** [comment]
raxml was never needed with kSNP4.  I included it initially because it was included in the Docker file for the previous version, and I thought that implied that it was useful (since I don't think raxml was required for 3.1 either).  Given the interest in a minimal package and how easy it is to install, I figured I'd remove it to save space and avoid the impression that it was a requirement.

It is easy to re-add if anyone wants it, including re-adding it to the Docker file.

I'm running the container manual test on my fork as we speak (didn't realize I was updating in public! :) )

**Comment by kissake** [comment]
Re: README.md at the package level, great catch, thanks!

**Comment by erinyoung** [comment]
Perfect!

Once the test is successful, I'll merge this PR and send to Quay and Dockerhub

**Review by erinyoung — APPROVED** [review]
All checks pass. Thank you for the PR. 

I'm going to
1. merge this PR into master
2. push to dockerhub and quay
3. using this tag as the 'lastest'

**Comment by erinyoung** [comment]
You can check the status of sending to dockerhub and quay at https://github.com/StaPH-B/docker-builds/actions/runs/3395298477

**Pull request merged**
